File: iso27001-phoenix/lib/iso27001_phoenix_web/controllers/v1/health_controller.ex
defmodule Iso27001PhoenixWeb.V1.HealthController do
  @moduledoc """
  A.17: Liveness, readiness, and detailed health endpoints.
  """
  use Phoenix.Controller, formats: [:json]
  import Plug.Conn

  alias Iso27001Phoenix.Repo
  alias Iso27001Phoenix.Guardian
  alias Iso27001Phoenix.Infrastructure.Telemetry.{ErrorBudget, QualityScore}

  @start_time System.monotonic_time(:second)

  @doc "GET /api/v1/health/live — no dependency checks (A.17)"
  def live(conn, _params) do
    json(conn, %{status: "ok"})
  end

  @doc "GET /api/v1/health/ready — DB ping required (A.17)"
  def ready(conn, _params) do
    db_status =
      try do
        Repo.query!("SELECT 1")
        "ok"
      rescue
        _ -> "error"
      end

    if db_status == "ok" do
      json(conn, %{status: "ok", checks: %{database: "ok"}})
    else
      conn
      |> put_status(503)
      |> json(%{status: "degraded", checks: %{database: "error"}})
    end
  end

  @doc "GET /api/v1/health/detailed — admin only (A.17)"
  def detailed(conn, _params) do
    with :ok <- require_admin(conn) do
      budget  = ErrorBudget.snapshot()
      score   = QualityScore.calculate(availability: budget.budget_remaining_pct / 100.0)
      uptime  = System.monotonic_time(:second) - @start_time

      json(conn, %{
        status:                      "ok",
        uptime_seconds:              uptime,
        error_budget_remaining_pct:  budget.budget_remaining_pct,
        error_budget_consumed_pct:   budget.budget_consumed_pct,
        quality_score:               Float.round(QualityScore.composite(score), 4),
        quality_passes_gate:         QualityScore.passes_gate?(score),
        slo_alerts:                  QualityScore.slo_alert(),
        components: %{
          database: db_check()
        }
      })
    end
  end

  defp require_admin(conn) do
    user = Guardian.Plug.current_resource(conn)

    if user && user.role == "admin" do
      :ok
    else
      conn
      |> put_status(403)
      |> json(%{error: %{code: "FORBIDDEN", message: "Admin role required"}})
      |> halt()
    end
  end

  defp db_check do
    try do
      Repo.query!("SELECT 1")
      "ok"
    rescue
      _ -> "error"
    end
  end
end

File: iso27001-nestjs/src/api/v1/health.controller.ts
import { Controller, Get, UseGuards } from '@nestjs/common';
import { InjectDataSource } from '@nestjs/typeorm';
import { DataSource } from 'typeorm';
import { JwtAuthGuard } from '../../core/guards/jwt-auth.guard';
import { RolesGuard } from '../../core/guards/roles.guard';
import { Roles } from '../../core/decorators/roles.decorator';
import { Public } from '../../core/decorators/public.decorator';
import { errorBudget } from '../../infrastructure/telemetry/error-budget.tracker';
import {
  QualityScoreCalculator,
} from '../../infrastructure/telemetry/quality-score.calculator';
import { getMetricsText, SLO_P95_LATENCY_MS, SLO_P99_LATENCY_MS } from '../../core/metrics/prometheus.metrics';

/**
 * A.17: Health check endpoints.
 * GET /health/live   — liveness (no dependencies)
 * GET /health/ready  — readiness (DB + cache)
 * GET /health/detail — full telemetry, admin-only
 * GET /metrics       — Prometheus metrics
 */
@Controller()
@UseGuards(JwtAuthGuard, RolesGuard)
export class HealthController {
  constructor(
    @InjectDataSource() private readonly dataSource: DataSource,
  ) {}

  @Public()
  @Get('health/live')
  liveness(): { status: string; timestamp: string } {
    return { status: 'ok', timestamp: new Date().toISOString() };
  }

  @Public()
  @Get('health/ready')
  async readiness(): Promise<{ status: string; checks: Record<string, unknown>; timestamp: string }> {
    const checks: Record<string, unknown> = {};
    let overall = true;

    try {
      const start = process.hrtime.bigint();
      await this.dataSource.query('SELECT 1');
      const latencyMs = Number(process.hrtime.bigint() - start) / 1_000_000;
      checks['database'] = { status: 'ok', latency_ms: Math.round(latencyMs * 100) / 100 };
    } catch (err) {
      checks['database'] = { status: 'error', detail: (err as Error).message };
      overall = false;
    }

    return {
      status: overall ? 'ok' : 'degraded',
      checks,
      timestamp: new Date().toISOString(),
    };
  }

  @Get('health/detail')
  @Roles('admin')
  detailedHealth(): Record<string, unknown> {
    const snapshot = errorBudget.snapshot();

    const calculator = new QualityScoreCalculator(
      SLO_P95_LATENCY_MS,
      500.0,
      SLO_P99_LATENCY_MS,
    );

    const score = calculator.calculate({
      authChecksPassed: snapshot.totalRequests - snapshot.failedRequests,
      authChecksTotal: snapshot.totalRequests,
      auditEventsRecorded: snapshot.totalRequests,
      auditEventsExpected: snapshot.totalRequests,
      availability: snapshot.observedAvailability,
      logsWithCorrelationId: snapshot.totalRequests,
      totalLogs: snapshot.totalRequests,
    });

    const alert = calculator.sloAlert(
      snapshot.failedRequests,
      snapshot.clientErrors,
      snapshot.totalRequests,
    );

    return {
      status: 'ok',
      timestamp: new Date().toISOString(),
      error_budget: {
        sla_target: snapshot.slaTarget,
        total_requests: snapshot.totalRequests,
        failed_requests: snapshot.failedRequests,
        client_errors: snapshot.clientErrors,
        observed_availability: snapshot.observedAvailability,
        budget_consumed_pct: snapshot.budgetConsumedPct,
        budget_exhausted: snapshot.budgetExhausted,
      },
      slo_alerts: {
        any_breach: alert.anyBreach,
        p95_latency_breached: alert.p95LatencyBreached,
        p99_latency_breached: alert.p99LatencyBreached,
        error_rate_breached: alert.errorRateBreached,
        client_error_spike: alert.clientErrorSpike,
      },
      quality_score: score,
    };
  }

  @Public()
  @Get('metrics')
  metrics(): string {
    return getMetricsText();
  }
}
